File: leetcode/dia 041/jumpII.js
var jump = function(nums) {
    let dp = {}, a;
    function rec(i){
        if (i === nums.length - 1) return 0;
        if (i in dp) return dp[i];
        dp[i] = Infinity;
        if (i + nums[i] < nums.length){
            for (let j = 0; j < nums[i]; j++){
                a = 1 + rec(i + j + 1);
                if (a < dp[i]) dp[i] = a;
            }
        }
        else{
            for (let j = 0; j < nums.length - i - 1; j++){
                a = 1 + rec(i + j + 1);
                if (a < dp[i]) dp[i] = a;
            }
        }
        return dp[i];
    }
    return rec(0);
};

File: leetcode/dia 041/jumpII.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        dp = {}
        def rec(i):
            if i == len(nums) - 1:
                return 0
            if i in dp:
                return dp[i]
            dp[i] = float("inf")
            if i + nums[i] < len(nums):
                for j in range(nums[i]):
                    dp[i] = min(dp[i], 1 + rec(i + j + 1))
            else:
                for j in range(len(nums) - i - 1):
                    dp[i] = min(dp[i], 1 + rec(i + j + 1))
            return dp[i]
        return rec(0)